File: Programming Paradigms/Segundo Parcial/Modelo Parcial 1/Tema1.pl
% Segundo Parcial - Paradigmas de Progrmación
% Turno 1 – Programación Lógica (Prolog)

%tipo_hamburguesa(CódigoTipoBurguer,Descripción,PesoEnGramosPorUnidad,Prec
 ioPorUnidad,PorcDescuento,ListadoIngredientes). 
tipo_hamburguesa('bur01','hamburguesa simple',200,1400,10,['pan de 
hamburguesa','medallón de carne','mayonesa']). 
tipo_hamburguesa('bur02','hamburguesa doble',300,2000,25,['pan de 
hamburguesa','medallón de carne','medallón de 
carne','chédar','mayonesa','kétchup']). 
tipo_hamburguesa('bur03','hamburguesa completa',350,2500,0,['pan de 
hamburguesa','medallón de carne','medallón de 
carne','chédar','beicon','mayonesa','kétchup']). 
tipo_hamburguesa('bur04','hamburguesa pollo',220,1800,20,['pan de 
hamburguesa','medallón de pollo','mayonesa']). 


% orden_elaboración(CódigoOrden,CódigoTipoBurguer, CódigoOperario 
%fecha(Día,Mes,Año),CantidadTotalUnidadesElaboradas,TiempoTotalEnMinutos). 
orden_elaboración('ord01','bur02','ope001',fecha(14,10,2024),150,75). 
orden_elaboración('ord02','bur03','ope001',fecha(14,10,2024),180,225). 
orden_elaboración('ord03','bur01','ope003',fecha(15,10,2024),120,180). 
orden_elaboración('ord04','bur04','ope002',fecha(15,10,2024),200,150). 


%operario(CódigoOperario,Nombre,Apellido). 
operario('ope001','María','Méndez'). 
operario('ope002','Rocío','González'). 
operario('ope003','Camila','Juárez'). 


% 1. Implementar una regla que permita calcular el precio total de una orden de elaboración,
% teniendo en cuenta el precio del tipo de hamburguesa correspondiente y la cantidad
% solicitada. En caso de que el tipo de hamburguesa tenga un descuento, este se aplica
% al total. 
% Nombre sugerido: precio_total_orden/2

% 2. Implementar una regla que reciba como parámetro el código de un operario y devuelva
% la cantidad total de hamburguesas que elaboró, considerando todas las órdenes que
% tenga asignadas. 
% Nombre sugerido: cantidad_total_operario/2

% 3. Implementar una regla que reciba una lista de códigos de orden de elaboración y devuelva
% la lista de nombres de los operarios que participaron en esas órdenes.
% Nombre sugerido: operarios_en_ordenes/2

File: Programming Paradigms/Segundo Parcial/Modelo Parcial 1/Tema1.hs
{- Segundo Parcial - Paradigmas de Programación
Turno 1 – Programación Funcional (Haskell)

1. Implementar una función que reciba como parámetro el código de un tipo de alfajor y
retorne el precio de venta unitario correspondiente, según los datos de la tabla 1: Tipos
de alfajores. En caso de recibirse un código de tipo de alfajor que no figure en la tabla,
la función deberá devolver 0.0. Utilizar expresión case of.
Nombre sugerido de la función: funcion1.

2. Implementar una función que reciba como parámetro una lista de códigos de tipos de
alfajores. La función deberá devolver la sumatoria total de los precios de venta de cada
uno de los tipos de alfajores cuyos códigos figuren en la lista recibida como parámetro.
Reutilizar la función del punto 1).
Nombre sugerido de la función: funcion2.

3. Implementar una función que reciba 3 parámetros: una lista de precios, un valor de
referencia min, y un valor de referencia max.
La función deberá devolver una lista con los precios de la lista recibida como primer
parámetro, tales que sean mayores o iguales al valor de referencia min y menores o
iguales al valor de referencia max. Utilizar lista por comprensión.
Nombre sugerido de la función: funcion3.
-}

funcion1 codigo 
	| codigo == 1 = "Paracetamol"
	| codigo == 2 = "Ibuprofeno"
	| codigo == 3 = "Salbutamol"
	| otherwise = "Inexistente" 

funcion2 _ [] = []
funcion2 ref (x:xs) 
		    | x > ref   = x : funcion2 ref xs
    		    | otherwise = funcion2 ref xs



funcion3 [] = 0
funcion3 (x:xs)
    | x >= 10 && x <= 50 = x + funcion3 xs
    | otherwise          = funcion3 xs
